### ISSUE TITLE
Deep Neural Network

### ISSUE THREAD

**Issue by ameer-alwadiya** [ameer-alwadiya/Practrical-Python#1]
Hey there,

I designed a deep learning network without using off-the-shelf libraries such as TensorFlow. However, when I try to apply an example like image classification, the loss doesn't decrease significantly; it only slightly decreases. I believe the problem is not from choosing the right parameters, but there is something wrong in the code that I can't find.

Here is the code:

class NeuralNetwork:
    def __init__(self, layer_dims, activation='relu'):
        self.layer_dims = layer_dims
        self.num_layers = len(layer_dims)
        self.activation = activation
        self.parameters = self.initialize_parameters()

    def initialize_parameters(self):
        np.random.seed(1)
        parameters = {}

        for l in range(1, self.num_layers):
            parameters['W' + str(l)] = np.random.randn(layer_dims[l], layer_dims[l-1]) * 0.01
            parameters['b' + str(l)] = np.zeros((layer_dims[l], 1))

        return parameters

    def forward_propagation(self, X):
        cache = {"A0": X}
        
        for l in range(1, self.num_layers):
            W = self.parameters["W" + str(l)]
            b = self.parameters["b" + str(l)]
            Z = np.dot(W, cache["A" + str(l - 1)]) + b
            
            # Use the selected activation function for hidden layers except for the last layer.
            if l < self.num_layers - 1:
                if self.activation == 'relu':
                    A = self.relu(Z)
                elif self.activation == 'sigmoid':
                    A = self.sigmoid(Z)
                else:
                    raise ValueError("Invalid activation function. Please choose 'relu' or 'sigmoid'.")
            else:
                A = self.sigmoid(Z)

            cache["Z" + str(l)] = Z
            cache["A" + str(l)] = A

        return cache["A" + str(self.num_layers - 1)], cache

    def compute_cost(self, Y_hat, Y):
        m = Y.shape[1]
        cost = (-1/m) * (np.dot(Y, np.log(Y_hat).T) + np.dot((1-Y), np.log(1-Y_hat).T))
                         
        return cost
        
    def backward_propagation(self, Y, cache):
        gradients = {}
        m = Y.shape[1]
        L = self.num_layers - 1

        dA = - (np.divide(Y, cache["A" + str(L)]) - np.divide(1 - Y, 1 - cache["A" + str(L)]))
        dZ = dA * self.sigmoid_derivative(cache["A" + str(L)])

        for l in range(L, 0, -1):
            gradients["dW" + str(l)] = np.dot(dZ, cache["A" + str(l - 1)].T) / m
            gradients["db" + str(l)] = np.sum(dZ, axis=1, keepdims=True) / m

            if l > 1:
                dA = np.dot(self.parameters["W" + str(l)].T, dZ)
                if self.activation == 'relu':
                    dZ = dA * self.relu_derivative(cache["Z" + str(l - 1)])
                elif self.activation == 'sigmoid':
                    dZ = dA * self.sigmoid_derivative(cache["Z" + str(l - 1)])
                else:
                    raise ValueError("Invalid activation function. Please choose 'relu' or 'sigmoid'.")

        return gradients

    def update_parameters(self, gradients, learning_rate):
                         
        for l in range(1, self.num_layers):
            self.parameters["W" + str(l)] -= learning_rate * gradients["dW" + str(l)]
            self.parameters["b" + str(l)] -= learning_rate * gradients["db" + str(l)]

    def train(self, X, Y, num_epochs=1000, learning_rate=0.01):
        costs = []
                         
        for epoch in range(num_epochs):
            Y_hat, cache = self.forward_propagation(X)
            cost = self.compute_cost(Y_hat, Y)
            gradients = self.backward_propagation(Y, cache)
            self.update_parameters(gradients, learning_rate)

            if epoch % 100 == 0:
                print(f"Epoch {epoch}, Cost: {cost}")
            costs.append(cost)

        return costs

    def predict(self, X):
        Y_hat, _ = self.forward_propagation(X)
        predictions = (Y_hat > 0.5).astype(int)
        return predictions
     
    def sigmoid(self, Z):
        return 1 / (1 + np.exp(-Z))

    def sigmoid_derivative(self, A):
        return A * (1 - A)
    
    def relu(self, Z):
        return np.maximum(0, Z)
    
    def relu_derivative(self, A):
        return 1 * (A > 0)
